### ISSUE TITLE
Performance Report: April 2022

### ISSUE THREAD

**Issue by maxyinger** [swissspidy/web-stories-wp#11263]
# Performance Report April 2022

This is a performance report covering progress in the editor from prior to performance work starting to the current state of editor performance. This is comparing branches main at hash `1d5c7228e` to branch `perf-sidebar` from [latest perf PR](https://github.com/GoogleForCreators/web-stories-wp/pull/11261)

## Environment

Interaction tests run on a fully specced 2018 13" Macbook Pro on [this test story](https://stories-new-wordpress-amp.pantheonsite.io/wp-admin/post.php?post=22410&action=edit#page=%25227dcbdd4c-186b-4cfe-80a7-a59c0b59c311%2522) which consists of 1 page with 50 square shape elements, each with a bounce animation, media pane open and layers panel expanded.

Each test interaction will document the main thread of 4 different builds:
- `1d5c7228e` React development build
- `1d5c7228e` React production build
- `perf-sidebar` React development build
- `perf-sidebar` React production build

Development builds are used to compare progress with last performance report #10158, whereas production builds are used to document approximate user experience progress. We are omitting the React flame graphs for now since the main focus here is user facing performance differences rather than how the interactions became more performant internally. Happy to document and add in React flame graphs adhoc if requested though.

| Branch: 1d5c7228e | Branch:  perf-sidebar |
| ---- | ---- |
| <img width="1440" alt="old_env" src="https://user-images.githubusercontent.com/35983235/163270262-02c6f588-2e55-4aac-aa50-5469d161ad4a.png"> |  <img width="1439" alt="new_env" src="https://user-images.githubusercontent.com/35983235/163270279-13cf3f45-446b-43e2-99bc-0ed1cd14cda9.png"> |

## TODO
 - [x] update selection interaction tests
 - [x] drag to update element position tests (drag start & end)
 - [x] reorder element layer interaction test 
 - [x] ~possible performance documentation on canvas generation cache improvements~ this is hard to clearly document, but with the new setup, we are hitting the cache more which should save us some ~1s calls to convert a story page to a canvs 
 
 ## TLDR
The most applicable metric to gauge user interaction is the time the interaction took in the react production build, because that's the build shipped to our users. Here's a small table highlighting the current performance increases for those interactions. Each interaction is elaborated on below with comments in this thread. 

Make note that we are officially under 100ms for all of these which is the bar Google sets for the [Interaction section of the RAIL model](https://web.dev/rail/#focus-on-the-user):

> 0 - 100ms
> Respond to user actions within this time window and users feel like the result is immediate. Any longer, and the connection between action and reaction is broken.
 
 | Interaction | Old Production Time | Current Production Time | Rough Performance Increase | Currently Under 100ms |
 | ---- | ---- | ---- | ---- | ---- |
 | Update Selection | 110ms | 35ms | 3.1x | ✅ |
 | Update Element Position - Drag Start | 50ms | 30ms | 1.7x | ✅ |
 | Update Element Position - Drag End | 195ms | 95ms | 2.0x | ✅ |
 | Reorder Layer Position | 120ms | 30ms | 4.0x | ✅ |

---
 
Fixes #11120

**Comment by littlemilkstudio** [comment]
# Selection Update
Here's the interaction we're recording:
https://user-images.githubusercontent.com/35983235/163421161-d03134d9-8ce1-4f77-b393-da6e59f73cab.mp4

**Note**
This video was taken on the `perf-sidebar` branch with react in the development build. Recorded the video on a 16" macbook pro m1 which is why the performance appears to be faster than the `perf-sidebar` dev build screenshot below (which was recorded on a 13" macbook pro with intel chip)

## React Development Build
| 1d5c7228e | perf-sidebar |
| ---- | ---- |
| ~320ms | ~70ms |
| <img width="482" alt="old_dev_selection" src="https://user-images.githubusercontent.com/35983235/163421977-13658afc-f0ef-41f1-ab80-f9c14b4df76d.png"> | <img width="456" alt="new_dev_selection" src="https://user-images.githubusercontent.com/35983235/163422009-6f02e270-755e-4367-afa5-b2e6d8456239.png"> |


## React Production Build
| 1d5c7228e | perf-sidebar |
| ---- | ---- |
| ~110ms | ~35ms |
| <img width="506" alt="old_build_selection" src="https://user-images.githubusercontent.com/35983235/163422352-87afd136-8033-4002-9f47-c802901dd34f.png"> |  <img width="507" alt="new_build_selection" src="https://user-images.githubusercontent.com/35983235/163422385-c0f358ef-ef27-47ec-9ac5-7a73609ae1f7.png"> |

**Comment by littlemilkstudio** [comment]
# Drag To Reposition Element

**Recording of Interaction**

https://user-images.githubusercontent.com/35983235/163423716-329e056e-9b1a-4749-b90b-5c10387f72bc.mp4

**Note**
This video was taken on the `perf-sidebar` branch with react in the development build. Recorded the video on a 16" macbook pro m1 which is why the performance appears to be faster than the `perf-sidebar` dev build screenshot below (which was recorded on a 13" macbook pro with intel chip)

## React Development Build

**Drag Start**
| 1d5c7228e | perf-sidebar |
| ---- | ---- |
| ~150ms | ~55ms |
| <img width="545" alt="old_dev_drag_start" src="https://user-images.githubusercontent.com/35983235/163424115-5de46be7-ae3b-4a87-b4a4-f292c8a87b77.png"> |  <img width="468" alt="new_dev_drag_start" src="https://user-images.githubusercontent.com/35983235/163424166-123963b1-6170-4c4a-b2c2-dd0e047e6f5e.png"> |

**Drag End**
| 1d5c7228e | perf-sidebar |
| ---- | ---- |
| ~585ms | ~140ms |
| <img width="545" alt="old_dev_drag_end" src="https://user-images.githubusercontent.com/35983235/163424377-5df9c32d-4167-445e-be88-ccdafeb743a2.png"> |  <img width="467" alt="new_dev_drag_end" src="https://user-images.githubusercontent.com/35983235/163424401-9941f7a6-a627-48ea-a970-87e88b5ceb10.png"> |

## React Production Build

**Drag Start**
| 1d5c7228e | perf-sidebar |
| ---- | ---- |
| ~50ms | ~30ms |
| <img width="507" alt="old_build_drag_start" src="https://user-images.githubusercontent.com/35983235/163424751-862c77ef-3267-4d20-8090-53210f3abb46.png"> |  <img width="415" alt="new_build_drag_start" src="https://user-images.githubusercontent.com/35983235/163424790-654e933f-3ee9-4fa7-a250-8d1f17feab1b.png"> |

**Drag End**
| 1d5c7228e | perf-sidebar |
| ---- | ---- |
| ~195ms | ~95ms |
| <img width="505" alt="old_build_drag_end" src="https://user-images.githubusercontent.com/35983235/163425096-7af598af-1fe9-47be-9249-08bd8dbe0ce6.png"> | <img width="457" alt="new_build_drag_end" src="https://user-images.githubusercontent.com/35983235/163425123-756867c2-3f87-433b-b965-97dfd1a9b927.png"> |

**Comment by littlemilkstudio** [comment]
# Reorder Element Layer

**Recording of Interaction**
https://user-images.githubusercontent.com/35983235/163426028-56d1e544-1710-46e1-a9f6-9e783b6aa84c.mp4

**Note**
This video was taken on the `perf-sidebar` branch with react in the development build. Recorded the video on a 16" macbook pro m1 which is why the performance appears to be faster than the `perf-sidebar` dev build screenshot below (which was recorded on a 13" macbook pro with intel chip)

## React Development Build

| 1d5c7228e | perf-sidebar |
| ---- | ---- |
| ~410ms | ~65ms |
| <img width="507" alt="old_dev_layer_reorder" src="https://user-images.githubusercontent.com/35983235/163426268-2c6fdf43-35b9-4cfb-8b64-40acdd47e8dc.png"> |  <img width="468" alt="new_dev_layer_reorder" src="https://user-images.githubusercontent.com/35983235/163426321-d83286cc-7a80-4f0f-9076-78652ab8d246.png"> |

## React Production Build

| 1d5c7228e | perf-sidebar |
| ---- | ---- |
| ~120ms | ~30ms |
| <img width="506" alt="old_build_layer_reorder" src="https://user-images.githubusercontent.com/35983235/163426539-8d6a69ed-cad2-4f99-b0f8-f31f4be9d35a.png"> |  <img width="458" alt="new_build_layer_reorder" src="https://user-images.githubusercontent.com/35983235/163426579-83526aac-d0cd-4f29-9cc7-17757ed95db6.png"> |
